File: Labs/Lab 1/q3.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

int min_energy(const std::string& s, char c = 'a') {
    if (s.length() == 1) {
        return (s[0] == c) ? 0 : 1;
    }

    int mid = s.length() / 2;
    int left_count = std::count(s.begin(), s.begin() + mid, c);
    int right_count = std::count(s.begin() + mid, s.end(), c);

    return std::min(mid - left_count + min_energy(s.substr(mid), static_cast<char>(c + 1)),
                    mid - right_count + min_energy(s.substr(0, mid), static_cast<char>(c + 1)));
}

int main() {
    int no;
    std::cin >> no;

    std::vector<int> l;
    for (int i = 0; i < no; i++) {
        int n;
        std::cin >> n;
        std::string s;
        std::cin >> s;
        int energy = min_energy(s);
        l.push_back((energy <= n) ? energy : -1);
    }

    for (int energy : l) {
        std::cout << energy << std::endl;
    }

    return 0;
}

File: Labs/Lab 1/q3.py
# Python code funderstanding the minimum energy required to convert a string to a string with all characters in alphabetical order
# and the Logic

def min_energy(s, char='a'):
    if len(s) == 1:
        return 0 if s[0] == char else 1

    mid = len(s) // 2
    cnt_left = s[:mid].count(char)
    cnt_right = s[mid:].count(char)

    return min(mid - cnt_left + min_energy(s[mid:], chr(ord(char) + 1)),
               mid - cnt_right + min_energy(s[:mid], chr(ord(char) + 1)))

l = []
t = int(input())
for _ in range(t):
    n = int(input())
    s = input()
    energy = min_energy(s)
    l.append(energy if energy <= n else -1)


print(*l, sep='\n')